Extension/src/Debugger/configurationProvider.ts

```diff
@@ -252,6 +232,64 @@ class CppConfigurationProvider implements vscode.DebugConfigurationProvider {
         return config && config.type ? config : null;
     }
 
+    private convertEnvFileToEnvironmentMap(config: vscode.DebugConfiguration, folder: vscode.WorkspaceFolder): void {
+        if (config.envFile) {
+            // replace ${env:???} variables
+            let envFilePath: string = util.resolveVariables(config.envFile, null);
+
+            try {
+                if (folder && folder.uri && folder.uri.fsPath) {
+                    // Try to replace ${workspaceFolder} or ${workspaceRoot}
+                    envFilePath = envFilePath.replace(/(\${workspaceFolder}|\${workspaceRoot})/g, folder.uri.fsPath);
+                }
+
+                const parsedFile: ParsedEnvironmentFile = ParsedEnvironmentFile.CreateFromFile(envFilePath, config["environment"]);
+
+                // show error message if single lines cannot get parsed
+                if (parsedFile.Warning) {
+                    CppConfigurationProvider.showFileWarningAsync(parsedFile.Warning, config.envFile);
+                }
+
+                config.environment = parsedFile.Env;
+                
+                delete config.envFile;
+            } catch (e) {
+                throw new Error(`Failed to use envFile. Reason: ${e.message}`);
+            }            
+        }
+    }
+
+    private resolveSourceFileMapVariables(config: vscode.DebugConfiguration): void {
+        let messages: string[] = [];
+        if (config.sourceFileMap) {
+            for (const sourceFileMapKey of Object.keys(config.sourceFileMap)) {
+                const sourceFileMapValue: string = config.sourceFileMap[sourceFileMapKey];
+
+                // TODO: pass config.environment as initial environment to resolveVariables when it is { key: value } instead of { "key": key, "value": value }
```

Comment:
should this be in the method above? this is talking about environment but you are in sourcefilemap

Reply:
No. This is for using config.environment variables in sourceFileMap.

E.g.
environment: {
   "abc": "efg"
},
sourceFileMap: {
   ${env:abc}: ...
}